contracts/contract/src/main.rs

```diff
@@ -0,0 +1,601 @@
+    init_events();
+
+    events::record_event_dictionary(Event::ChangeEventsMode(ChangeEventsMode {
+        events_mode: events_mode_u8,
+    }));
+}
+
+pub fn upgrade(name: &str) {
+    let entry_points = generate_entry_points();
+
+    let package_key_name = &format!("{PREFIX_CEP18}_{PREFIX_CONTRACT_PACKAGE_NAME}_{name}");
+    let contract_key_name = &format!("{PREFIX_CEP18}_{PREFIX_CONTRACT_NAME}_{name}");
+
+    let old_contract_package_hash = match runtime::get_key(package_key_name)
+        .unwrap_or_revert_with(Cep18Error::FailedToGetOldPackageKey)
+    {
+        Key::Hash(contract_hash) => contract_hash,
+        Key::AddressableEntity(EntityAddr::SmartContract(contract_hash)) => contract_hash,
+        Key::SmartContract(package_hash) => package_hash,
+        _ => revert(Cep18Error::MissingPackageHashForUpgrade),
+    };
+    let contract_package_hash = PackageHash::new(old_contract_package_hash);
+
+    let previous_contract_hash = match runtime::get_key(contract_key_name)
+        .unwrap_or_revert_with(Cep18Error::FailedToGetOldContractHashKey)
+    {
+        Key::Hash(contract_hash) => contract_hash,
+        Key::AddressableEntity(EntityAddr::SmartContract(contract_hash)) => contract_hash,
+        _ => revert(Cep18Error::MissingContractHashForUpgrade),
+    };
+    let converted_previous_contract_hash = AddressableEntityHash::new(previous_contract_hash);
+
+    let events_mode = get_optional_named_arg_with_user_errors::<u8>(
+        ARG_EVENTS_MODE,
+        Cep18Error::InvalidEventsMode,
+    );
+
+    let version_value_uref = get_uref_with_user_errors(
+        &format!("{PREFIX_CEP18}_{PREFIX_CONTRACT_VERSION}_{name}"),
+        Cep18Error::MissingVersionContractKey,
+        Cep18Error::InvalidVersionContractKey,
+    );
+
+    let version_value: String = read(version_value_uref)
+        .unwrap_or_default()
+        .unwrap_or_default();
+
+    // If stored version is a non empty string (and not a u32), it means it is already a 2.0
+    // version, do not add message topics then, as already set when installed
+    let message_topics: BTreeMap<String, MessageTopicOperation> = if !version_value.is_empty() {
+        BTreeMap::new()
+    } else {
+        BTreeMap::from([(ARG_EVENTS.to_string(), MessageTopicOperation::Add)])
+    };
+
+    let mut named_keys = NamedKeys::new();
+    named_keys.insert(ARG_CONDOR.to_string(), storage::new_uref(ARG_CONDOR).into());
+
+    let (contract_hash, contract_version) = storage::add_contract_version(
+        contract_package_hash.into(),
+        entry_points,
+        named_keys,
+        message_topics,
+    );
+
+    storage::disable_contract_version(
+        contract_package_hash.into(),
+        converted_previous_contract_hash.into(),
+    )
+    .unwrap_or_revert_with(Cep18Error::FailedToDisableContractVersion);
+
+    // migrate old ContractPackageHash as PackageHash so it's stored in a uniform format with the
```

Comment:
wrong comment?

Reply:
Seems to me almost correct 

contract_package_hash is a PackageHash

add_contract_version uses add_contract_version

key uses PackageHash as key

I changed it to 
```
    // migrate old ContractPackageHash as PackageHash so it's stored in a uniform format with the
    // `add_contract_version` implementation
    
    ```

Reply:
so far, it's stored as a ContractPackage key, not Package. Also, I wonder if this `put_key` is changing or creating a new value? or since we're in upgrade, it was there already.
<img width="1073" alt="image" src="https://github.com/user-attachments/assets/af59a474-556d-4b6d-a454-9f1736113ae6" />